[Test] Define `broken`/`skip` keyword argument for `@test`

giordano opened the pull request:
They make it less tedious having tests failing consistently only under specific
conditions.

giordano committed:
[Test] Define `broken`/`skip` keyword argument for `@test`

They make it less tedious having tests failing consistently only under specific
conditions.

giordano commented:
Side note, I feel like `@test_broken` and `@test_skip` should have been
```julia
@test ... broken=true
@test ... skip=true
```
but that ship has sailed already.  Maybe good for v2.0?

oscardssmith commented:
If we want that for 2.0, we should add the keyword version now so the ecosystem has plenty of time to transition.

giordano commented:
If there is consensus about adding `broken`/`skip` keywords to `@test`, I can look into that (not very soon though)

giordano commented on stdlib/Test/src/Test.jl:
````diff
@@ -406,6 +408,36 @@ macro test_broken(ex, kws...)
+```jldoctest
+julia> @test_broken_if true 2 + 2 == 5
+Test Broken
+  Brokenped: 2 + 2 == 5
````
I have absolutely no idea where this "Brokenped" was coming from :flushed:

giordano commented:
I've updated the pull request to define the `broken`/`skip` keyword arguments for `@test`.  I'll later update replace some occurrences of `@test_broken`/`@test_skip` in the test suite where the new form makes more sense

giordano committed:
[Test] Use new `@test ... broken=cond` in some tests

giordano commented on test/complex.jl:
```diff
@@ -1088,16 +1088,8 @@ end
-            @test isequal(complex(one(T)) / complex(T(Inf), T(-Inf)), complex(zero(T), zero(T)))
-            @test isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T)))
-        end
+        @test isequal(complex(one(T)) / complex(T(Inf), T(-Inf)), complex(zero(T), zero(T))) broken=(T==Float64)
+        @test isequal(complex(one(T)) / complex(T(-Inf), T(Inf)), complex(-zero(T), -zero(T))) broken=(T in (Float32, Float64))
```
This is the kind of simplification that the new form enables

staticfloat replied:
this is amazing

giordano commented:
Bump :slightly_smiling_face:

vtjnash commented on stdlib/Test/src/Test.jl:
```diff
@@ -368,12 +370,73 @@ Test Passed
-    :(do_test($result, $orig_ex))
+    testres = :(Broken(:skipped, $orig_ex))
+
+    return quote
+        if $(length(broken) > 0 && esc(broken[1]))
+            # code to call do_test with execution result and original expr
+            do_broken_test($result, $orig_ex)
+        elseif $(length(skip) > 0 && esc(skip[1]))
+            record(get_testset(), $testres)
+        else
+            do_test($result, $orig_ex)
+        end
+    end
```
```suggestion
    testres = :(do_test($result, $orig_ex))
    if length(skip) > 0
        testres = :($(esc(skip[1])) ? Broken(:skipped, $orig_ex) : $testres)
    end
    if length(broken) > 0
        testres = :($(esc(broken[1])) ? do_broken_test($result, $orig_ex) : $testres)
    end
    return testres
```

staticfloat replied:
I think we want to avoid calling `do_test()` at all if `length(skip) > 0`.

giordano replied:
How about
```suggestion
    return quote
        if $(length(skip) > 0 && esc(skip[1]))
            record(get_testset(), Broken(:skipped, $orig_ex))
        else
            _do = $(length(broken) > 0 && esc(broken[1])) ? do_broken_test : do_test
            _do($result, $orig_ex)
        end
    end
```
which has a single copy of `result` as suggested by Jeff?

miguelraz commented:
Triage asks to add a NEWS.md item to show the `broken=...` (the tests under `test/complex.jl` would be nice).

JeffBezanson commented on stdlib/Test/src/Test.jl:
```diff
@@ -368,12 +370,73 @@ Test Passed
+        elseif $(length(skip) > 0 && esc(skip[1]))
+            record(get_testset(), $testres)
+        else
+            do_test($result, $orig_ex)
```
Should avoid putting multiple copies of `result` in the output when possible.

JeffBezanson commented:
Very nice feature!!

giordano commented:
I believe I addressed all comments from triage:

* simplified the expression returned by `@test`, with a single copy of `result`
* added a `NEWS.md` entry showing a simplified test from `test/complex.jl`

giordano commented:
Bump :slightly_smiling_face: The failure on macOS is in Downloads, and looks unrelated (I think)

giordano commented:
Good to merge? :smiley:
